g3doc/quick_reference.md

```diff
@@ -1404,6 +1404,29 @@ Ops in this section are only available if `HWY_TARGET != HWY_SCALAR`:
     encryption: `SubBytes(ShiftRows(state)) ^ round_key`. This matches x86
     AES-NI. The latency is independent of the input values.
 
+*   `V`: `u8` \
+    <code>V **AESInverseRound**(V state, V round_key)</code>: one round of AES
```

Comment:
> I'm not sure we want to provide this one. Is it something you are using? If not, perhaps we can just adopt the x86 semantics like AESRound does, because the penalty for other platforms is relatively minor (extra Xor + generating a zero constant) whereas emulating AESInverseRound on x86 is expensive (6..8 cycle latency for IMC, and only every other cycle on SKX).
> 
> My current thinking is that it would be nice to expose x86 AESDEC as AESRoundInv or -Dec or Inv/DecRound (instead of the current EquivInvCipherRound, which is confusing because its inverse is just AESRound), and also LastRound.
> 
> But explicitly not the other algorithm, nor IMC - probably better to nudge people away from using it because it's super slow on x86 (no 256 nor 512-bit versions) and we aim to smooth out performance cliffs as much as possible. What do you think?

The Equivalent Inverse Cipher algorithm described on page 25 of https://nvlpubs.nist.gov/nistpubs/fips/nist.fips.197.pdf actually applies the InvMixColumns operation to each of the expanded round keys in its key expansion operation, and the InvMixColumns operation would still likely be needed by decryption code that uses the AESRoundInv operation if AESRoundInv has the same semantics as x86 AESDEC.

GCC, Clang, and MSVC will also apply common subexpression elimination optimizations to _mm_aesimc_si128 calls whenever possible when optimizations are enabled.

Another issue is that the InvMixColumns operation has extra overhead on PPC8/PPC9/PPC10 as PPC8/PPC9/PPC10 does not have a single instruction for the InvMixColumns operation that x86 AES-NI, AArch64 Crypto, and SVE Crypto have. The InvMixColumns operation can be emulated through a ```_mm256_aesdec_epi128(_mm256_aesenclast_epi128(state, _mm256_setzero_si128()), _mm256_setzero_si128())``` operation on PPC8/PPC9/PPC10.

It is more efficient to implement InvMixColumns for 256-bit vectors on AVX3_DL/AVX3_ZEN using ```_mm256_aesdec_epi128(_mm256_aesenclast_epi128(state, _mm256_setzero_si128()), _mm256_setzero_si128())``` instead of splitting the 256-bit vector to 128-bit vectors, carrying out 2 _mm_aesimc_si128 operations, and combining the results of the 2 _mm_aesimc_si128 operations back into a 256-bit vector.

Reply:
Good point that IMC is still useful for implementing the key schedule, makes sense to expose it then.

I think most applications would be OK with lower key agility=some overhead when changing keys caused by slower emulation of IMC. Intel's IMC is also relatively slow.

Thanks for making the change (removing the alternative to the equivalent inverse cipher) :) This avoids the 'which one should I use' confusion for users.